xarray/core/parallelcompat.py

```diff
@@ -0,0 +1,265 @@
+
+    loaded_entrypoints = {
+        entrypoint.name: entrypoint.load() for entrypoint in entrypoints
+    }
+
+    # TODO will this work if dask is not installed? We don't want to instantiate the chunkmanager if its not available
+    available_chunkmanagers = {
+        name: chunkmanager()
+        for name, chunkmanager in (loaded_entrypoints | EXAMPLE_CHUNKMANAGERS).items()
+        if chunkmanager.available
+    }
+    return available_chunkmanagers
+
+
+def guess_chunkmanager_name(manager: Optional[str]) -> str:
+    chunkmanagers = list_chunkmanagers()
+
+    if manager is None:
+        if len(chunkmanagers) == 1:
+            # use the only option available
+            manager = next(iter(chunkmanagers.keys()))
+        else:
+            # default to trying to use dask
+            manager = "dask"
+
+    return manager
+
+
+def guess_chunkmanager(manager: Optional[str]) -> "ChunkManagerEntrypoint":
+    """
+    Get namespace of chunk-handling methods, guessing from what's available.
+
+    If the name of a specific ChunkManager is given (e.g. "dask"), then use that.
+    Else use whatever is installed, defaulting to dask if there are multiple options.
+    """
+
+    chunkmanagers = list_chunkmanagers()
+    manager = guess_chunkmanager_name(manager)
+
+    if isinstance(manager, str):
+        if manager not in chunkmanagers:
+            raise ValueError(
+                f"unrecognized chunk manager {manager} - must be one of: {list(chunkmanagers)}"
+            )
+
+        return chunkmanagers[manager]
+    else:
+        # TODO should we accept type[ChunkManagerEntrypoint] too?
+        raise TypeError("manager must be a string")
+
+
+def get_chunked_array_type(*args) -> "ChunkManagerEntrypoint":
+    """
+    Detects which parallel backend should be used for given set of arrays.
+
+    Also checks that all arrays are of same chunking type (i.e. not a mix of cubed and dask).
+    """
+
+    # TODO this list is probably redundant with something inside xarray.apply_ufunc
+    ALLOWED_NON_CHUNKED_TYPES = {int, float, np.ndarray}
+
+    chunked_arrays = [
+        a
+        for a in args
+        if is_chunked_array(a) and type(a) not in ALLOWED_NON_CHUNKED_TYPES
+    ]
+
+    # Asserts all arrays are the same type (or numpy etc.)
+    chunked_array_types = {type(a) for a in chunked_arrays}
+    if len(chunked_array_types) > 1:
+        raise TypeError(
+            f"Mixing chunked array types is not supported, but received multiple types: {chunked_array_types}"
+        )
+    elif len(chunked_array_types) == 0:
+        raise TypeError("Expected a chunked array but none were found")
+
+    # iterate over defined chunk managers, seeing if each recognises this array type
+    chunked_arr = chunked_arrays[0]
+    chunkmanagers = list_chunkmanagers()
+    for chunkmanager in chunkmanagers.values():
+        if chunkmanager.is_chunked_array(chunked_arr):
+            return chunkmanager
+
+    raise ChunkManagerNotFoundError(
```

Comment:
```suggestion
    raise TypeError(
```

?

Reply:
Resolved in https://github.com/pydata/xarray/pull/7019/commits/2c18df61d8a2676d1a02a5475d70b54484a4b668